File: C++/7-Reverse Integer.cpp
// Runtime: 4 ms, faster than 92.23% of C++ online submissions for Reverse Integer.
// Memory Usage: 8.2 MB, less than 68.28% of C++ online submissions for Reverse Integer.
class Solution {
public:
    int reverse(int x) {
        int res = 0;
        while (x != 0) {
            if (res > INT_MAX / 10 || res < INT_MIN / 10) return 0;
            res = res * 10 + x % 10;
            x /= 10;
        }
        return res;
    }
};


// official
class Solution {
public:
    int reverse(int x) {
        int rev = 0;
        while (x != 0) {
            int pop = x % 10;
            x /= 10;
            if (rev > INT_MAX/10 || (rev == INT_MAX / 10 && pop > 7)) return 0;
            if (rev < INT_MIN/10 || (rev == INT_MIN / 10 && pop < -8)) return 0;
            rev = rev * 10 + pop;
        }
        return rev;
    }
};

File: Python3.x/7-Reverse Integer.py
# 暴力
# Runtime: 36 ms, faster than 92.74% of Python3 online submissions for Reverse Integer.
# Memory Usage: 13.2 MB, less than 74.27% of Python3 online submissions for Reverse Integer.
class Solution:
    def reverse(self, x: int) -> int:
        tmp = str(x)
        if tmp[0] == '-':
            result = int(tmp[0] + tmp[1:][::-1])
            if result > 2 ** 31 or result < -2 ** 31:
                return 0
            else:
                return result
        else:
            result = int(tmp[::-1])
            if result > 2 ** 31 or result < -2 ** 31:
                return 0
            else:
                return result


# 精简后
# Runtime: 28 ms, faster than 99.54% of Python3 online submissions for Reverse Integer.
# Memory Usage: 13.3 MB, less than 53.02% of Python3 online submissions for Reverse Integer.
class Solution:
    def reverse(self, x: int) -> int:
        tmp = str(x)
        if tmp[0] == '-':
            result = int(tmp[0] + tmp[1:][::-1])
            return 0 if result > 2 ** 31 or result < -2 ** 31 else result
        else:
            result = int(tmp[::-1])
            return 0 if result > 2 ** 31 or result < -2 ** 31 else result


# Runtime: 36 ms, faster than 92.74% of Python3 online submissions for Reverse Integer.
# Memory Usage: 13.2 MB, less than 74.27% of Python3 online submissions for Reverse Integer.
class Solution:
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        res = int(str(abs(x))[::-1]) if x > 0 else -int(str(abs(x))[::-1])
        return res if -2 ** 31 - 1 < res < 2 ** 31 else 0
